### PR TITLE
MAINT: Move z/(z - 1) transform case for hyp2f1 into Cython

### PR DESCRIPTION
**Pull request by steppi**
#### Reference issue

This PR closes no issues.
#### What does this implement/fix?

This PR is part of the ongoing work to rewrite scipy's Gaussian hypergeometric function implementation for complex numbers in Cython one small piece at a time. This PR handles the cases where the `z/(z - 1)` transform [(DLMF 15.8.1)](https://dlmf.nist.gov/15.8#i) is used. Where their convergence regions overlap, the `z/(z -1)` transform is preferred over the maclaurin series for `real(z) < 0` because it avoids the oscillation of terms in the maclaurin series for `real(z) < 0` that causes numerical issues due to cancellation.

The Fortran implementation in `scipy/special/specfun/specfun.f` used this tranform when `|z| < 1` and `real(z) < 0`, but the region has been expanded to include points with `|z| < 1.1`. The Fortran implementation uses the `1/z` transform up until the very edge of its region of convergence where it has numerical difficulties.

Comments have been updated to no longer refer cryptically to "The original Fortran implementation" but instead to cite Shanjie Zhang and Jianming Jin's implementation explictly.
#### Additional information
Benchmarking using the default test cases from `scipy.special._precompute.hyp2f1_data` in the relevant regions in the complex plane, and for parameters that are not special cases that are not touched by this PR show comparable results between the new implementation and the old implementation for `real(z) < 0` and `|z| <= 1`. The following table shows the quartiles for relative error over the 212,800 test cases in this region. The new Cython implementation shows improvement despite being nearly identical. I haven't attempted to account for this discrepancy.

|   method      |   min |         25% |         50% |         75% |     max |
|-------------|-------|-------------|-------------|-------------|---------|
| new-cython  |     0 | 3.9172e-16  | 8.0316e-16  | 2.80373e-15 | 0.0055  |
| old-fortran |     0 | 4.16948e-16 | 1.02367e-15 | 5.7686e-13  | 1.33226 |

For `real(z) < 0 ` and `1 < |z| < 1.1` the improvements due to expanding the region where the `z/(z - 1)` transform is used are substantial. This next table shows quartiles for relative error over the 33,600 test cases in this region.

| method      |   min |         25% |         50% |         75% |         max |
|-------------|-------|-------------|-------------|-------------|-------------|
| new-cython  |     0 | 7.31658e-16 | 1.33853e-15 | 5.52456e-15 | 0.102492    |
| old-fortran |     0 | 1.10815e-09 | 1.88101e-08 | 6.73397     | 1.51509e+75 |

### CODE REVIEW

**Comment by steppi** [comment]
@person142, once again, the actual diff outside of test cases is small.

File: scipy/special/tests/test_hyp2f1.py [hunk 1]
```diff
@@ -1578,7 +1818,260 @@ def test_region5(self, hyp2f1_test_case):
         assert 1 < abs(z) < 1.1 and abs(1 - z) >= 0.9 and z.real >= 0
         assert_allclose(hyp2f1(a, b, c, z), expected, rtol=rtol)
 
-    # Marked as slow so it won't run by default.
```
Comment (h-vetinari) [inline scipy/special/tests/test_hyp2f1.py:1581]: Seems this comment was removed unintentionally (as the following test still has `@pytest.mark.slow`)?

**Review by person142 — APPROVED** [review]
Looks good; LMK if CI failures should be ignored.

File: scipy/special/tests/test_hyp2f1.py (hunk 1 above)
Comment (steppi) [reply scipy/special/tests/test_hyp2f1.py:1581]: Sorry. Strictly speaking I should have mentioned this in the PR message. I think this comment is no longer necessary. It started its life explaining why the test was marked slow even though the test was actually very fast. As test cases have been added, the test actually has become slow and the comment has been whittled down to just this. The shortened comment is actually not true. Slow tests do run by default, but on CI they only run for the full jobs.

**Comment by steppi** [comment]
> Looks good; LMK if CI failures should be ignored.

Thanks. I think the only failures are from when git clones were taking an excessive amount of time over the holidays. I've rebased and force pushed so the tests will run again. It would be good to see the results on the mac tests just in case small differences in floating point arithmetic cause failures that don't occur on windows or linux.

**Comment by steppi** [comment]
@person142 all of the failures look unrelated. I think it's safe to ignore themi

**Comment by person142** [comment]
Thanks, in it goes then.

**Pull request merged**